### ISSUE TITLE
Reinforcement Learning Bots

### ISSUE THREAD

**Issue by garbear** [garbear/xbmc#89]
Reinforcement learning is showing great promise for gameplay. Yesterday, AlphaZero destroyed the best Chess AI using a network trained from scratch in four hours ([article](https://www.chess.com/news/view/google-s-alphazero-destroys-stockfish-in-100-game-match), [paper](https://arxiv.org/pdf/1712.01815.pdf)). This issue explores the idea of creating bots that use reinforcement learning to play games in Kodi.

# Theory

## Concepts

Emulation can be abstracted as a global *state* that evolves by applying an *action* and running the frame:

* State consists of the video and audio output, as well as memory regions (RAM, SRAM, Real-time clock)
* Action is the combined state of all controllers before the frame.

Reinforcement learning adds two concepts to the state and action of emulation: *reward* and *policy*.

* Reward is used to train the function approximator that infers an action from the observed state. The reward can come from sniffing RAM, such as the achievements at http://retroachievements.org, or reading a number from video memory using OCR.
* Policy is used by the agent to choose the next move. The goal of reinforcement learning is to choose the policy that maximizes the reward.

These four concepts are given the following variables:

![State: S](http://latex.codecogs.com/gif.latex?State%5Cequiv%7BS%7D)
![Reward: R](http://latex.codecogs.com/gif.latex?Reward%5Cequiv%7BR%7D)
![Policy: pi](http://latex.codecogs.com/gif.latex?Policy%5Cequiv%7B%5Cpi%7D)
![Action: A](http://latex.codecogs.com/gif.latex?Action%5Cequiv%7BA%7D)

These four concepts are all the gameplay data we need to train a reinforcement learner.

Emulation occurs at discrete timesteps, so every timestep has its own instance of these variables:

![S == S_t](http://latex.codecogs.com/gif.latex?S%5Cequiv%7BS_t%7D)
![R == R_t](http://latex.codecogs.com/gif.latex?R%5Cequiv%7BR_t%7D)
![pi == pi_t](http://latex.codecogs.com/gif.latex?%5Cpi%5Cequiv%7B%5Cpi_t%7D)
![A == A_t](http://latex.codecogs.com/gif.latex?A%5Cequiv%7BA_t%7D)

The emulation history is therefore a timeseries of a tuple containing these four variables.

## Timesteps

Timesteps occur by applying a set of functions to the emulation variables:

![Update()](http://latex.codecogs.com/gif.latex?Update%28%29)
* The *Update()* function takes the previous state, along with the most recent values of the other variables, and produces a new state

![Reward()](http://latex.codecogs.com/gif.latex?Reward%28%29)
* The *Reward()* function takes the previous reward, along with the most recent values of the other variables, and produces a new reward

![Strategize()](http://latex.codecogs.com/gif.latex?Strategize%28%29)
* The *Strategize()* function takes the previous policy, along with the most recent values of the other variables, and produces a new policy

![Act()](http://latex.codecogs.com/gif.latex?Act%28%29)
* The *Act()* function takes the previous action, along with the most recent values of the other variables, and produces a new action

## Model

The model consists of the initial conditions, as well as the values at each timestep.

The initial condition of all variables is the empty set:

![S_0 = 0](http://latex.codecogs.com/gif.latex?S_0%3D%5Cemptyset)
![R_0 = 0](http://latex.codecogs.com/gif.latex?R_0%3D%5Cemptyset)
![pi_0 = 0](http://latex.codecogs.com/gif.latex?%5Cpi_0%3D%5Cemptyset)
![A_0 = 0](http://latex.codecogs.com/gif.latex?A_0%3D%5Cemptyset)

The variables then evolve by applying the functions in sequence:

![S_t+1 = Update(S_t, R_t, pi_t, A_t)](http://latex.codecogs.com/gif.latex?S_%7Bt%2B1%7D%3DUpdate%28S_t%2CR_t%2C%5Cpi_t%2CA_t%29)
![R_t+1 = Reward(S_t+1, R_t, pi_t, A_t)](http://latex.codecogs.com/gif.latex?R_%7Bt%2B1%7D%3DReward%28S_%7Bt%2B1%7D%2CR_t%2C%5Cpi_t%2CA_t%29)
![pi_t+1 = Strategize(S_t+1, R_t+1, pi_t, A_t)](http://latex.codecogs.com/gif.latex?%5Cpi_%7Bt%2B1%7D%3DStrategize%28S_%7Bt%2B1%7D%2CR_%7Bt%2B1%7D%2C%5Cpi_t%2CA_t%29)
![A_t+1 = Act(S_t+1, R_t+1, pi_t+1, A_t)](http://latex.codecogs.com/gif.latex?A_%7Bt%2B1%7D%3DAct%28S_%7Bt%2B1%7D%2CR_%7Bt%2B1%7D%2C%5Cpi_%7Bt%2B1%7D%2CA_t%29)

**Comment by LipkeGu** [comment]
As far i understand, you mean to develop a "virtual" player which is available in each game / rom?

**Comment by garbear** [comment]
right. so far the math in this issue just describes the data we need to gather to make this happen. then it can be uploaded to the cloud for training, and depending on the state of embedded tensorflow, inference can be run locally or in the cloud if we get netplay support.
